text/0401-coercions.md

```diff
@@ -319,20 +319,23 @@ descriptions are equivalent.
 Casting is indicated by the `as` keyword. A cast `e as U` is valid if one of the
 following holds:
 
-* `e` has type `T` and `T` coerces to `U`;
-
-* `e` has type `*T` and `U` is `*U_0` (i.e., between any raw pointers);
-
-* `e` has type `*T` and `U` is `uint` , or vice versa;
-
-* `e` has type `T` and `T` and `U` are any numeric types;
-
-* `e` is a C-like enum and `U` is any integer type, `bool`;
-
-* `e` has type `T` and `T == u8` and `U == char`;
-
-* `e` has type `T` and `T == &[V, ..n]` or `T == &V` and `U == *const V`, and
-  similarly for the mutable variants to either `*const V` or `*mut V`.
+ * `e` has type `T` and `T` coerces to `U`; *coercion-cast*
+ * `e` has type `*T`, `U` is `*U_0`, and either `U_0: Sized` or
+    unsize_kind(`T`) = unsize_kind(`U_0`); *ptr-ptr-cast*
+ * `e` has type `*T` and `U` is a numeric type, while `T: Sized`; *ptr-addr-cast*
+ * `e` is an integer and `U` is `*U_0`, while `U_0: Sized`; *addr-ptr-cast*
+ * `e` has type `T` and `T` and `U` are any numeric types; *numeric-cast*
+ * `e` is a C-like enum and `U` is an integer type; *enum-cast*
+ * `e` has type `bool` or `char` and `U` is an integer; *prim-int-cast*
+ * `e` has type `u8` and `U` is `char`; *u8-char-cast*
+ * `e` has type `&[T; n]` and `U` is `*const T`; *array-ptr-cast*
+ * `e` is a function pointer type and `U` has type `*T`,
+   while `T: Sized`; *fptr-ptr-cast*
+ * `e` is a function pointer type and `U` is an integer; *fptr-addr-cast*
+
+where `&.T` and `*T` are references of either mutability,
+and where unsize_kind(`T`) is the kind of the unsize info
+in `T` - a vtable or a length (or `()` if `T: Sized`).
```

Comment:
There are potentially two ways to interpret this "definition" for `unsize_kind` ... either `unsize_kind` says whether the type is a trait, or a slice, or a sized type (i.e. there are only *three* possible values that it returns), or `unsize_kind` says the same thing but also includes *which* trait (i.e. which vtable) for the first case, and *which* particular length for the second case (i.e. that its a much larger range of potential values).

I infer that the intention is the latter; i.e. that it is not legal e.g. to cast from one trait-object to another with an unrelated vtable via `as`, nor is it legal to cast from one array to another with a different length.  But it still would be nice if the text didn't require any thinking on this point.

Anyway, I wouldn't block this change based on this detail; we can always come up with a later PR to try to clarify this later.